hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStoreMutationObserver.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.recordstore;
+
+import com.hazelcast.map.impl.record.Record;
+import com.hazelcast.nio.serialization.Data;
+
+/**
+ * Interface for observing {@link RecordStore} mutations
+ *
+ * @param <R> The type of the records in the observed {@link RecordStore}
+ */
+public interface RecordStoreMutationObserver<R extends Record> {
```

Comment:
can't we use event-journal mechanism to observe mutations instead of introducing this `RecordStoreMutationObserver`?

Reply:
From @blazember:
> finally, I decided to not share the same abstraction for the event journal and the merkle tree update calls
they're only mostly the same, there are some exceptions when they need to behave differently, like `RecordStore.reset()`, `RecordStore.putRecord()`
it would be a bit odd to distinguish between the cases in the callbacks
also, one captures events, the other captures state changes, they're different if we consider replication

Reply:
How i see all these is: data is changing and there are some pluggable mechanisms which want to observe these changes. Maybe subject to another improvement, but it seems we should provide an abstraction to do all these, otherwise complexity is increasing. Event journal, merkle trees or map-store... all seems somewhat similar.

Reply:
I agree these should share the same abstraction. Ideally, these should be hooked on the `Storage` instead of the `RecordStore`, since that is the closest place to the data mutation and the code flow there is more straight than it is in the `RecordStore`. But, it would require some change to the `Storage` interface (see the methods in the comment above), so I decided not to do this.

I moved `EventJournal` behind the `RecordStoreMutationObserver` abstraction. This comes with an increased number of observer methods. Some as a NOP implementation in the EventJournal observer. Let's see what you guys think.